proxying.go

```diff
@@ -0,0 +1,74 @@
+package fronted
+
+import (
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/getlantern/uuid"
+)
+
+// DialTimeout dials out using masquerading
+func DialTimeout(proxyHost string, timeout time.Duration) (net.Conn, error) {
+	_d, ok := _instance.Get(timeout)
+	if !ok {
+		return nil, dialError("Timeout waiting for masquerade")
+	}
+	d := _d.(*direct)
+	// TODO: apply timeout to dial here too
+	conn, masqueradeGood, err := d.dial()
+	return &proxyingConn{Conn: conn, proxyHost: proxyHost, masqueradeGood: masqueradeGood}, err
+}
+
+type dialError string
+
+func (e dialError) Error() string {
+	return string(e)
+}
+
+func (e dialError) Timeout() bool {
+	return true
+}
+
+func (e dialError) Temporary() bool {
+	return true
+}
+
+type proxyingConn struct {
+	net.Conn
+	proxyHost      string
+	masqueradeGood func(bool) bool
+}
+
+// OnRequest implements the proxy.RequestAware interface to prepare the domain-
+// fronted request.
+func (conn *proxyingConn) OnRequest(req *http.Request) {
+	if req.URL.Scheme == "" {
+		// HTTPS requests had their scheme stripped, add it back
+		req.URL.Scheme = "https"
+	}
+	// Store original URL for domain-fronting
+	req.Header.Set("X-Ddf-Url", req.URL.String())
```

Comment:
Isn't this available as part of the path in the request?

Reply:
No, because we're domain-fronting everything via the same CloudFront URL (`http://d100fjyl3713ch.cloudfront.net/`). Also, in addition to the path, we need the original origin and the protocol scheme (http vs https). Having the full original URL gives us everything we need.